File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/validation/comparator/DefaultValueComparator.java
package org.skellig.teststep.processing.validation.comparator;

import java.util.ArrayList;
import java.util.Collection;

public class DefaultValueComparator implements ValueComparator {

    private final Collection<ValueComparator> comparators;

    private DefaultValueComparator(Collection<ValueComparator> comparators) {
        this.comparators = comparators;
    }

    @Override
    public boolean compare(Object actualValue, Object expectedValue) {
        return comparators.stream()
                .filter(comparator -> comparator.isApplicable(expectedValue))
                .anyMatch(comparator -> comparator.compare(actualValue, expectedValue));
    }

    @Override
    public boolean isApplicable(Object expectedValue) {
        return true;
    }

    public static class Builder {
        private Collection<ValueComparator> valueComparators;

        public Builder() {
            valueComparators = new ArrayList<>();
            withValueComparator(new ContainsValueComparator());
            withValueComparator(new RegexValueComparator());
        }

        public Builder withValueComparator(ValueComparator valueComparator) {
            valueComparators.add(valueComparator);
            return this;
        }

        public ValueComparator build() {
            withValueComparator(new EqualsValueComparator());
            return new DefaultValueComparator(valueComparators);
        }
    }
}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/validation/comparator/DefaultValueComparator.kt
package org.skellig.teststep.processing.validation.comparator

class DefaultValueComparator(private val comparators: MutableMap<String, ValueComparator>) : ValueComparator {

    override fun compare(comparator: String, args: Array<Any?>, actualValue: Any?): Boolean {
        return comparators[comparator]?.compare(comparator, args, actualValue) ?: false
    }

    override fun getName(): String = ""

    class Builder {
        private val valueComparators: MutableMap<String, ValueComparator> = mutableMapOf()

        init {
            withValueComparator(ContainsValueComparator())
            withValueComparator(MatchValueComparator())
            withValueComparator(LessThanValueComparator())
            withValueComparator(LessOrEqualsValueComparator())
            withValueComparator(MoreThanValueComparator())
            withValueComparator(MoreOrEqualsThanComparator())
            withValueComparator(DateTimeValueComparator())
        }

        fun withValueComparator(valueComparator: ValueComparator) = apply {
            valueComparators[valueComparator.getName()] = valueComparator
        }

        fun build(): ValueComparator {
            val defaultValueComparator = DefaultValueComparator(valueComparators)
            withValueComparator(NotValueComparator(defaultValueComparator))
            withValueComparator(EqualsValueComparator())
            return defaultValueComparator
        }
    }
}